pandas/core/arrays/base.py

```diff
@@ -18,19 +20,20 @@ class ExtensionArray(object):
 
     * __getitem__
     * __len__
+    * __iter__
```

Comment:
@jorisvandenbossche we removed this earlier since collections implementing `__len__` and `__getitem__` are iterable. But (strangely?) `isinstance(arr, collections.Iterable)` is False, so these aren't considered "list-like" unless they actually implement `__iter__`.

Reply:
But do we use this somewhere? (the expectation that isinstance(arr, collections.Iterable) is True) 

As the docs https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable says this is not a correct way to test that something is iterable, exactly for the reason you just mentioned.

Anyway, if we add it, I would add a default implementation (`return iter(self)` ?) and not require subclasses to provide one manually.

Reply:
> But do we use this somewhere?

Yes, sorry. We use it in `Series.__init__`, via `is_list_like(data)`.

Will add the default implementation.

Reply:
Wait, `iter(self)` will cause infinite recursion, and we can't assume anything about how the data is stored. That's unfortunate.

Reply:
I guess we could fall back to

```python
for i in range(len(self)):
   yield self[i]
```

Reply:
> Yes, sorry. We use it in Series.__init__, via is_list_like(data).

Or we could fix `is_list_like` ..

> Wait, iter(self) will cause infinite recursion, and we can't assume anything about how the data is stored.

Why would it cause infinite recursion? 
With `iter(self)` we don't assume anything about how the data is stored, I don't know how iter exactly works under the hood, but I assume it just uses `__getitem__` ?

Reply:
> Or we could fix is_list_like 

Prior to https://github.com/pandas-dev/pandas/pull/17654 we did that, but it was fragile to things like `str` (the class, not an instance) being considered iterable. Not sure if changing it back is an improvement.

> Why would it cause infinite recursion?

```
def __iter__(self):
    return iter(self)
```

will call `self.__iter__`, which calls `self.__iter__`, ...

Reply:
ah, yes of course :-) 
Anywhow, the other way is a fine default implementation